File: DesktopTools/dotNET/DesktopToolApp/DesktopToolApp/Functions/DeviceMaintenance/PingTester.cs
using System;
using System.Threading.Tasks;
using static DesktopTool.FunctionMessage;

namespace DesktopTool
{
    internal class PingTester : ToolDoProcess
    {
        // PINGバイトを保持
        private Random RandomInst { get; set; }
        private byte[] PingBytes { get; set; }

        public PingTester(string menuItemName) : base(menuItemName) 
        {
            RandomInst = new Random();
            PingBytes = new byte[PingTestQuery.PingBytesSize];
        }

        protected override void InvokeProcessOnSubThread()
        {
            // 100バイトのランダムデータを生成
            RandomInst.NextBytes(PingBytes);

            // PINGテスト処理を実行
            Task task = Task.Run(PerformPingTest);
        }

        private void PerformPingTest()
        {
            // PINGテスト処理を実行
            new PingTestQuery().Inquiry(NotifyResponseQueryHandler, PingBytes);
        }

        private void NotifyResponseQueryHandler(PingTestQuery sender, bool success, string errorMessage)
        {
            if (success == false) {
                // 失敗時はログ出力
                TerminateCommand(false, errorMessage);
                return;
            }

            // PINGバイトの一致チェック
            for (int i = 0; i < PingBytes.Length; i++) {
                if (PingBytes[i] != sender.PingResponseBytes[i]) {
                    TerminateCommand(false, MSG_PING_TEST_INVALID_RESPONSE);
                    return;
                }
            }

            // 画面に制御を戻す
            TerminateCommand(true, string.Empty);
        }

        //
        // 終了処理
        //
        private void TerminateCommand(bool success, string message)
        {
            // 終了メッセージを画面表示／ログ出力
            if (message.Length > 0) {
                if (success) {
                    LogAndShowInfoMessage(message);
                } else {
                    LogAndShowErrorMessage(message);
                }
            }

            // 画面に制御を戻す
            PauseProcess(success);
        }
    }
}

File: DesktopTools/macOS/Functions/DeviceMaintenance/PingTester.m
//
//  PingTester.m
//  DesktopTool
//
//  Created by Makoto Morita on 2024/02/01.
//
#import "FunctionMessage.h"
#import "PingTester.h"
#import "PingTestQuery.h"

@interface PingTester () <PingTestQueryDelegate>
    // 上位クラスの参照を保持
    @property (nonatomic) id                            delegate;
    @property (nonatomic) PingTestQuery                *pingTestQuery;
    // PINGデータを保持
    @property (nonatomic) NSData                       *pingRequestData;

@end

@implementation PingTester

    - (instancetype)initWithDelegate:(id)delegate {
        self = [super initWithDelegate:delegate];
        if (self) {
            [self enableClickButtonDoProcess:false];
            [self setPingTestQuery:[[PingTestQuery alloc] initWithDelegate:self]];
        }
        return self;
    }

    - (void)PingTestQuery:(PingTestQuery *)pingTestQuery didUpdateState:(bool)available {
        if (available) {
            [self enableClickButtonDoProcess:true];
        }
    }

    - (void)invokeProcessOnSubQueue {
        // 100バイトのランダムデータを生成
        uint8_t pingBytes[100];
        [self generateRandom:pingBytes length:sizeof(pingBytes)];
        [self setPingRequestData:[[NSData alloc] initWithBytes:pingBytes length:sizeof(pingBytes)]];
        // PINGテスト処理を実行
        [[self pingTestQuery] inquiryWithData:[self pingRequestData]];
    }

    - (void)generateRandom:(uint8_t *)randomBytes length:(NSUInteger)length {
        for (int i = 0; i < length; i++) {
            // from 0 to 255
            randomBytes[i] = (unsigned char)arc4random_uniform(256);
        }
    }

    - (void)PingTestQuery:(PingTestQuery *)pingTestQuery didNotifyResponseQuery:(bool)success withErrorMessage:(NSString *)errorMessage {
        if (success == false) {
            [self terminateCommand:false withMessage:errorMessage];
            return;
        }
        // PINGバイトの一致チェック
        if ([[self pingRequestData] isEqualToData:[pingTestQuery pingResponseData]] == false) {
            [self terminateCommand:false withMessage:MSG_PING_TEST_INVALID_RESPONSE];
            return;
        }
        // 画面に制御を戻す
        [self terminateCommand:true withMessage:nil];
    }

#pragma mark - 終了処理

    - (void)terminateCommand:(bool)success withMessage:(NSString *)message {
        // 終了メッセージを画面表示／ログ出力
        if (success) {
            [self LogAndShowInfoMessage:message];
        } else {
            [self LogAndShowErrorMessage:message];
        }
        [self pauseProcess:success];
    }

@end
